### PR TITLE
[1822] Don't send courses marked 'new' to UCAS

### PR DESCRIPTION
**Pull request by faissalMT**
### Context
When UCAS requests course data we should not provide courses that have been marked 'new' as these haven't yet been published (and also have the potential to be discarded instead).

### Changes proposed in this pull request
This PR ensures that courses with a status of 'new' are not emitted by the v1 courses endpoint.  

### Guidance to review
This has the side effect of causing empty courses not to be sent, but this is correct behaviour. I had to correct tests that were dependent on the previous behaviour.  
This PR is a redo of #641 for merging into master.  

### Checklist

- [X] Make sure all information from the Trello card is in here
- [X] Attach to Trello card
- [X] Rebased master
- [X] Cleaned commit history
- [X] Tested by running locally

### CODE REVIEW

**Comment by faissalMT** [comment]
[Implement deleted-course behaviour in APIv1 for UCAS](https://trello.com/c/kNRLJZQy/1822-implement-deleted-course-behaviour-in-apiv1-for-ucas)

File: app/models/course.rb [hunk 1]
```diff
@@ -121,6 +121,12 @@ def new_draft_attributes
     end.order(:changed_at, :id)
   end
 
+  scope :not_new, -> do
```
Comment (timabell) [inline app/models/course.rb]: Ronseal, does what it says on the tin. :+1:

File: lib/tasks/govuk_lint.rake [hunk 2]
```diff
@@ -3,7 +3,7 @@ namespace :lint do
   task :ruby do
     puts 'Linting...'
     unless system 'bundle exec rubocop app config db lib spec Gemfile --format clang -a'
-      exit $?
+      exit $?.exitstatus
```
Comment (timabell) [inline lib/tasks/govuk_lint.rake]: I think the commit message should explain why this change was necessary

File: spec/requests/api/v1/course_spec.rb [hunk 3]
```diff
@@ -123,16 +122,22 @@ def get_course_codes_from_body(body)
...
-               course_code: "LAST1",
-               age: 10.minutes.ago,
-               provider: provider)
+        Timecop.freeze(10.minutes.ago) do
+          first_course = create(:course,
+                                course_code: 'LAST1',
+                                provider: provider)
+
+          create(:site_status, :published, course: first_course)
+        end
```
Comment (timabell) [inline spec/requests/api/v1/course_spec.rb]: Re: the diff on this file; it's not clear to me from the commit message why this patch on this file is so large. I'm sure it's a good improvement, but I can't tell from reading it without a long time getting into your head why it was needed or desired for this particular PR.

Generally I'd suggest either:

* explanation in the commit message as to why you needed/wanted to improve the file
* overall improvments to the spec that are not specific to the task or that are improvements to be built on should be pulled out into a separate base commit or maybe even a separate PR

File: spec/factories/site_statuses.rb [hunk 4]
```diff
@@ -22,6 +22,7 @@
     transient do
       any_vancancy { false }
       provider { build :provider }
+      age { nil }
```
Comment (defong) [inline spec/factories/site_statuses.rb]: whats the `age` for?

File: spec/requests/api/v1/course_spec.rb [hunk 5]
```diff
@@ -398,5 +410,31 @@ def get_next_courses(link, params = {})
...
+      let(:course2) { create(:course, study_mode: 'full_time', profpost_flag: 'postgraduate', program_type: 'higher_education_programme', provider: provider1) }
+
+      let(:status1) { create(:site_status, status: :new_status, course: course1, vac_status: 'full_time_vacancies', applications_accepted_from: Date.new(2019, 7, 23)) }
+      let(:status2) { create(:site_status, status: :new_status, course: course2, vac_status: 'full_time_vacancies', applications_accepted_from: Date.new(2019, 7, 24)) }
+      let(:status3) { create(:site_status, status: :running, course: course2, vac_status: 'full_time_vacancies', applications_accepted_from: Date.new(2019, 7, 24)) }
+
+      it 'does not send courses marked new' do
+        status1
+        status2
+        status3
```
Comment (timabell) [inline spec/requests/api/v1/course_spec.rb]: use `let!` instead of `let` to cause immediate execution of the block, this will eliminate the need for calling them at the start of the `it` block

File: spec/factories/site_statuses.rb (hunk 4 above)
Comment (faissalMT) [reply spec/factories/site_statuses.rb]: It's a remnant of a fix I was attempting, I ended up just going with TimeCop in the end.

File: app/models/course.rb [hunk 6]
```diff
@@ -121,6 +121,12 @@ def new_draft_attributes
...
+    includes(site_statuses: %i[site course])
+      .where
+      .not(SiteStatus.table_name => { status: SiteStatus.statuses[:new_status] })
```
Comment (defong) [inline app/models/course.rb]:
```suggestion
      .not(SiteStatus.table_name => { status: :new_status })
```

Can it not be like that?

File: spec/requests/api/v1/course_spec.rb (hunk 3 above)
Comment (faissalMT) [reply spec/requests/api/v1/course_spec.rb]: These changes were done because the tests break without it. In these tests the courses have no `site_statuses` and so this PR causes those courses not to show, since this is correct behaviour it means those courses must be given `site_statuses`. Since creating these associations updates certain things I've used TimeCop to make those guarantees.  
I've improved the commit message to clarify.

File: app/models/course.rb (hunk 6 above)
Comment (faissalMT) [reply app/models/course.rb]: https://stackoverflow.com/questions/36487900/rails-using-enum-in-a-where-query

File: spec/requests/api/v1/course_spec.rb (hunk 5 above)
Comment (defong) [reply spec/requests/api/v1/course_spec.rb]: or a before block, its confusing looking at

```ruby

it 'test something' do
do_lots_of_setup

do_actual_thing

expect(actual).to eq (do_lots_of_setup)

```
